pkg/machineproviders/providers/openshift/machine/v1beta1/mapping.go

```diff
@@ -82,14 +110,62 @@ func createBaseFailureDomainMapping(cpms *machinev1.ControlPlaneMachineSet, fail
 func createMachineMapping(ctx context.Context, logger logr.Logger, cl client.Client, cpms *machinev1.ControlPlaneMachineSet) (map[int32]failuredomain.FailureDomain, error) {
 	out := make(map[int32]failuredomain.FailureDomain)
 
-	// TODO: Use the CPMS selector to fetch Machines from the API. Extract the providerconfig from these to then get the
-	// failure domain information out of the Machine. Inspect the Machine name, if it ends with an integer, take that as
-	// its index and set the failure domain in the output map. Newest machines should take precedence when inferring the
-	// failure domain if multiple machines match an index.
-	// If the Machine name does not end in an integer (or it is out of bounds of the replicas expected), then ignore the
-	// Machine.
-	// To test "Newest machines should take precedence", the logic after fetching the machines should be in a helper
-	// function that can be unit tested separately.
+	selector, err := metav1.LabelSelectorAsSelector(&cpms.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert label selector to selector: %w", err)
+	}
+
+	machineList := &machinev1beta1.MachineList{}
+	if err := cl.List(ctx, machineList, &client.ListOptions{LabelSelector: selector}); err != nil {
+		return nil, fmt.Errorf("failed to list machines: %w", err)
+	}
+
+	// indexToMachine contains a mapping between the machine domain index in the newest machine
+	// for this particular index.
+	indexToMachine := make(map[int32]machinev1beta1.Machine)
+
+	for _, machine := range machineList.Items {
+		failureDomain, err := providerconfig.ExtractFailureDomainFromMachine(machine)
+		if err != nil {
+			return nil, fmt.Errorf("could not extract failure domain from machine %s: %w", machine.Name, err)
+		}
+
+		machineNameIndex, ok := parseMachineNameIndex(machine.Name)
+		if !ok {
+			// Ignore the machine as it doesn't contain an index in its name.
+			logger.V(4).Info(
+				"Ignoring machine in failure domain mapping with unexpected name",
+				"machine", machine.Name,
+			)
+
+			continue
+		}
+
+		if fd, ok := out[int32(machineNameIndex)]; ok && fd.String() != failureDomain.String() {
```

Comment:
We have a function for comparing these
```suggestion
		if fd, ok := out[int32(machineNameIndex)]; ok && !fs.Equal(failureDomain) {
```